### DISCUSSION TITLE
Creating a dart-only library with a flutter library on top of it

### DISCUSSION THREAD

**Discussion by GregoryConrad** [fzyzcjy/flutter_rust_bridge#819] [Q&A] [answered]
Hello!

Many packages out today have a dart-only base, and then a flutter package built on top of it (think riverpod). I was wondering how that would look with `flutter_rust_bridge`, if possible.

Here is what I was thinking of making:
1. Low level dart-only package wrapping around a rust library (should be easy enough, will _not_ support web)
2. Low level flutter package that depends on and re-exports dart-only package 1 and provides some flutter-only abstractions to make working with package 1 easier (will _not_ support web)
3. High level dart-only package that depends on package 1 that makes it easier to work with in state management solutions (_might_ support web, but not through rust)
4. High level flutter package that depends on packages 2 & 3 (essentially a combination of the 2). This one would be used by most end users (_might_ support web, but not through rust)

Is this sort of setup possible (with all rust bindings exclusively in package 1)? How would this look amongst all packages? Any other caveats I should be aware of?

I am mostly wondering how package packages 2 & 3 would need to look. 1 & 4 should be trivial enough (I think?).

**Comment by fzyzcjy** [comment, marked as answer]
> have a dart-only base, and then a flutter package built on top of it

Same in my internal app indeed. 

As for web - flutter_rust_bridge does support web, so guess will not have trouble about that.

> Is this sort of setup possible (with all rust bindings exclusively in package 1)?

Yes, works well in my internal app

For me, I combine 1+3 and 2+4, so there is only a dart package and a flutter package

**Reply by GregoryConrad** [reply]
Thanks for the swift response and input! I was wondering if there would be any specific caveats with a Flutter package importing a dart-only package with rust ffi (perhaps needing some specific configuration), but glad to hear that it sounds pretty simple. I'll report back here in case anything doesn't work as I would expect.

**Reply by GregoryConrad** [reply]
@fzyzcjy Ok! Ran into a couple questions. The dart examples in this repo use a `dylibPath`, which makes sense (unless there's a way to statically link?).

For my Dart library (package 1 above), I was going to make users pass in the dylib path; is the correct approach?

To build a Flutter library (package 2 above) on top of such a Dart library with the dylib path passed in, how do I supply the Dart library with the link to this `dylibPath`? I did not see that in the Flutter example. I have a feeling I am thinking about this wrong, so I would appreciate your input on the matter.

**Reply by fzyzcjy** [reply]
> I was going to make users pass in the dylib path; is the correct approach?

LGTM since we have no idea where it is - it can be anywhere. Or if you enforce users to put somewhere then can hardcode the path (e.g. suppose you put .so in the lib folder etc)

> how do I supply the Dart library with the link to this dylibPath? I did not see that in the Flutter example

The example does show this IIRC

**Reply by GregoryConrad** [reply]
Thanks again. I'll take another look at the the flutter example and see if I might have missed anything.

**Reply by fzyzcjy** [reply]
You are welcome!

**Reply by GregoryConrad** [reply]
Looks like I glossed over the `ffi.io.dart` file on my first passthrough because I figured it was generated, but that was in fact the missing piece of the puzzle and thus why I was confused! All starting to make sense now.

On a side note, I was wondering if there was any any reason you use `DynamicLibrary.process` for iOS and `DynamicLibrary.executable` for macOS in [`ffi.io.dart`](https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_example/with_flutter/lib/ffi.io.dart)? I did some digging to try to figure out the difference and ran into [this issue](https://github.com/dart-lang/sdk/issues/44856#issue-801271056), which suggests they are identical, but perhaps something didn't work in testing?

**Reply by fzyzcjy** [reply]
https://docs.flutter.dev/development/platform-integration/ios/c-interop

```
final DynamicLibrary nativeAddLib = Platform.isAndroid
    ? DynamicLibrary.open('libnative_add.so')
    : DynamicLibrary.process();
```

**Reply by fzyzcjy** [reply]
Btw feel free to PR to add this link to wherever you get confused

**Reply by GregoryConrad** [reply]
Hi! My project has been coming along well. Dart-only was working far easier than I expected (got it working first try!). I also think I might take a stab at a PR for #828 in a week or two (I've been picking up Rust very fast and it has become my favorite language by far).

I also had another question: I see that with enums, a freezed class is generated by default. When using a struct, a regular dart class is generated by default. Is it possible with the code gen as-is to instead generate freezed classes for regular structs? If not, I think I'd like to make a PR to add that functionality because I need it in several parts (currently wrapping a struct in an enum to get freezed support and it's less than ideal).

At first, I was thinking it would be nice to make all generated classes use `freezed`, but that would be a breaking change. Then, I was thinking it might be nice to instead opt-in to the functionality with something like `#[frb(freezed)]` or `#[frb_freezed]`, similar to the `#[frb(mirror(SomeExternalType))]` like you currently have. Thoughts on this? Any pointers on what files to look at in particular when making a PR (assuming this functionality doesn't already exist)?

**Reply by fzyzcjy** [reply]
> I might take a stab at a PR for https://github.com/fzyzcjy/flutter_rust_bridge/issues/828 in a week or two (I've been picking up Rust very fast and it has become my favorite language by far).

Looking forward! Btw I also love Rust very much ;)

> Is it possible with the code gen as-is to instead generate freezed classes for regular structs? If not, I think I'd like to make a PR to add that functionality because I need it in several parts (currently wrapping a struct in an enum to get freezed support and it's less than ideal).

Forget the details whether this is implemented, maybe check the doc a bit?

> #[frb(freezed)] 

That grammar LGTM

https://github.com/fzyzcjy/flutter_rust_bridge/issues/468
https://github.com/fzyzcjy/flutter_rust_bridge/pull/463#discussion_r876517135

**Reply by GregoryConrad** [reply]
Looks like I just missed it in the docs! Thanks. I'll see when I can get around to that bug fix PR.

**Reply by GregoryConrad** [reply]
Hi again! Sorry for keeping on bothering you. I finally finished most of the dart-only library wrapping around Rust. For reference, you can take a peek [here](https://github.com/GregoryConrad/mimir/tree/main/packages/mimir). It's still a big WIP so don't mind the README or any sloppy Rust code.

I am now getting confused on how to create the flutter library that relies upon the dart-only library. To start out, I am adding in macOS/iOS support and will work on the other platforms later on. [The documentation](https://cjycode.com/flutter_rust_bridge/integrate/ios_linking.html) deals with the Runner project itself, but no Runner project is available when creating a flutter package/library. For guidance, I looked toward flutter's `plugin_ffi` template, which has a simple `plugin_ffi.podspec` file.

Here's what I'm currently thinking: I have the project created by `cargo-xcode`, and I need to incorporate it somehow into the `podspec`. I believe you mentioned you have created a flutter library wrapping around some rust code; would you be willing to share the relevant parts of just the `podspec` file? I looked in this repo for an example and could not find one (no podspec files anywhere).

**Reply by fzyzcjy** [reply]
> would you be willing to share the relevant parts of just the podspec file? I looked in this repo for an example and could not find one (no podspec files anywhere).

Ah yes, it is used internally in my app so not in this repo yet. I personally have a whole python-based infra so not suitable for anyone without this infra indeed. The code is as follows. Some may not be needed (because in the old days I used C++ but later change to pure Rust so some may only be useful for c++).

```
Pod::Spec.new do |s|
...
  s.module_map = 'my_vision_utils_flutter_wrapper.modulemap' // forget whether this is needed or not needed
...
  s.pod_target_xcconfig = {
    'OTHER_LDFLAGS' => '-lvision_utils_rs', # my lib is named vision_utils_rs
  }

  s.static_framework = true
  s.public_header_files = 'Classes**/*.h'
  s.vendored_libraries = "**/*.a"

  s.script_phase = {
      :name => 'Compile Rust',
      :script => %q{env -i HOME="$HOME" PODS_TARGET_SRCROOT="$PODS_TARGET_SRCROOT" CONFIGURATION="$CONFIGURATION" /bin/bash -c 'source ~/.bash_profile && source ~/.bash_profile && cd $PODS_TARGET_SRCROOT/../symlink_python_mono && python3 -m python_mono.scripts.vision_utils ios build --profile $CONFIGURATION'},
      :execution_position => :before_compile,
      :shell_path => '/bin/sh'
   }
end
```

Btw the lib looks interesting, looking forward to it :)

**Reply by GregoryConrad** [reply]
Thank you so much! I'll see how it works with mine.

**Reply by fzyzcjy** [reply]
You are welcome!

**Reply by GregoryConrad** [reply]
Ended up needing to tweak what you gave for the podspec slightly and am running into a roadblock. [My SO post](https://stackoverflow.com/questions/74467824/creating-flutter-rust-ffi-plugin-with-an-xcframework-for-macos-ios) in case you had any ideas.

**Reply by GregoryConrad** [reply]
On a side note: it looks like creating an XCFramework from the rust static libraries seems to be the best way to go right now. In case anyone else comes across this discussion, here is a copy-paste of the build script I used to create the XCFramework (will need to be modified for other cases):
```bash
#!/bin/bash

# Build static libs
cd $( dirname -- "$0"; )
for TARGET in \
        aarch64-apple-ios x86_64-apple-ios aarch64-apple-ios-sim \
        x86_64-apple-darwin aarch64-apple-darwin
do
    rustup target add $TARGET
    cargo build -r --target=$TARGET
done

# Create XCFramework
FRAMEWORK="EmbeddedMilli.xcframework"
LIBNAME=libembedded_milli.a
mkdir mac-lipo ios-sim-lipo
IOS_SIM_LIPO=ios-sim-lipo/$LIBNAME
MAC_LIPO=mac-lipo/$LIBNAME
lipo -create -output $IOS_SIM_LIPO \
        target/aarch64-apple-ios-sim/release/$LIBNAME \
        target/x86_64-apple-ios/release/$LIBNAME
lipo -create -output $MAC_LIPO \
        target/aarch64-apple-darwin/release/$LIBNAME \
        target/x86_64-apple-darwin/release/$LIBNAME
xcodebuild -create-xcframework \
        -library $IOS_SIM_LIPO \
        -library $MAC_LIPO \
        -library target/aarch64-apple-ios/release/$LIBNAME \
        -output $FRAMEWORK
rm -rf ios-sim-lipo mac-lipo

# Copy XCFramework to macOS and iOS folders
for OS in macos ios
do
cp -r $FRAMEWORK ../../flutter_mimir/$OS/
done

# Delete the intermediary XCFramework
rm -rf $FRAMEWORK
```

**Reply by fzyzcjy** [reply]
Here is podspec for macos (a bit different from ios)

```
Pod::Spec.new do |s|
  s.module_map = 'my_vision_utils_flutter_wrapper.modulemap'
  s.pod_target_xcconfig = {
    'DEFINES_MODULE' => 'YES', 'VALID_ARCHS[sdk=iphonesimulator*]' => 'x86_64',
    'CLANG_CXX_LANGUAGE_STANDARD' => 'c++17',
    'OTHER_CPLUSPLUSFLAGS' => ["-fno-aligned-allocation", "$(OTHER_CFLAGS)"],
    'OTHER_LDFLAGS' => '-lvision_utils_rs',
    'USER_HEADER_SEARCH_PATHS' => '"${PROJECT_DIR}/.."/',
    "LIBRARY_SEARCH_PATHS" => '"${PROJECT_DIR}/.."/*',
  }
  s.static_framework = true
  s.public_header_files = 'Classes**/*.h'
  s.vendored_libraries =  'libvision_utils_rs.dylib' # NOTE, not sure whether related
  s.script_phase = {
      :name => 'Compile Rust',
      :script => %q{env -i HOME="$HOME" PODS_TARGET_SRCROOT="$PODS_TARGET_SRCROOT" CONFIGURATION="$CONFIGURATION" /bin/bash -c 'source ~/.bash_profile && cd $PODS_TARGET_SRCROOT/../symlink_python_mono && python3 -m python_mono.scripts.vision_utils macos build --profile $CONFIGURATION'},
      :execution_position => :before_compile,
      :shell_path => '/bin/sh'
   }
end
```

**Reply by GregoryConrad** [reply]
Finally got iOS, macOS, and Android working with my library (demo video below)! What a ride it has been. Windows and Linux are hopefully coming soon, and perhaps I will tackle Web in the future.

On another note: web support is an entirely different animal. Since I need `std` in Rust, I have two options:
- Use a libc for wasm written in JS, [which surprisingly has been done before](https://leaningtech.com/webvm-server-less-x86-virtual-machines-in-the-browser/)
  - Unfortunately, that implementation is not publicly available _yet_ (not sure if it ever will ever be, but I am hopeful)
- Compile down to WASM/WASI and use [wasmer-js](https://github.com/wasmerio/wasmer-js) for a WASI polyfill.

It looks like neither of these options have been explored much for `flutter_rust_bridge` (and it seems Flutter in general), at least based on what I read in the documentation, so I will report back later with what I find.

![Demo Video](https://media3.giphy.com/media/5CXp6KYJIyuhLApWNK/giphy.gif)

**Reply by fzyzcjy** [reply]
Congratulations!

**Reply by coder0107git** [reply]
@GregoryConrad It looks like Leaning Technologies (the company's website is the top link in your last post) has made their libc implementation opensource [here](https://github.com/leaningtech/cheerp-newlib). They also have instructions on how to build their version of libc++ and libc++abi [here](https://github.com/leaningtech/cheerp-compiler#build-libcxx-and-libcxxabi).

**Reply by dinbtechit** [reply]
@GregoryConrad I had the exact same question. I am trying to build a "Dart only" library and use that library to build a flutter library on top of it. 

I am glad to see that you got it successfully working on android, ios, and macOS.  Are you able to post an example git repo with some step-by-step instructions on how you got it working? It would be massively helpful. 🙏🙏

**Reply by GregoryConrad** [reply]
@dinbtechit I got it working with every platform so far but web (web is still a TODO--need to see if I can get it working with WASI/wasmer-js hopefully, and if not, then the libc @coder0107git linked). I frankly think I could write a novel on everything I learned through the process--it was a lot more in depth than I was hoping to go. You will want a repo structure and tooling very similar to what I have in mimir for your project.

You can see my working implementation at https://github.com/gregoryconrad/mimir. I even threw CI integration tests in there (for every platform but android--the android reasoning is a bit complicated) so I know it is constantly working for all supported platforms. Here's a quick summary:

Note: my solution relies on distributing the binaries separately alongside a GitHub release. this is necessary because pub.dev has a hard 100mb upload limit, which my binaries would not fit into. This complicates things _a lot_, but I still got it working! Also, you are not supposed to upload binaries to pub.dev anyways, so I would recommend going with my approach.

Another note: I also set up my project to allow cross-compiling to all platforms _from any platform_, which is super convenient (except iOS/macOS have to be built on a Mac. I am using one though so I didn't bother researching into that area more). Take a look at my CONTRIBUTING.md to see system requirements for building. Also see my GitHub ci workflows to figure out how to build/publish releases automatically.

## Apple platforms
After some research, it appeared that an XCFramework is the current recommended approach for distributing cross-platform/architecture libraries. I also use static libraries for iOS/macOS in the XCFramework since that seemed to be the best approach for the platforms. Thus, I made a build-apple.sh script that creates said XCFramework and zips it up. iOS/macOS use the *exact same* podspec file because of this, and thus I think it is the right approach. Take a look at packages/flutter_mimir/ios/flutter_mimir.podspec to see how it works!

## Android
I compile all my android code into an .aar (which is actually just a .zip with a certain dir structure) which I then distribute thru GitHub release. Android was a bit unique since it directly references the file from my GitHub releases using an "ivy repository". For the other platforms, I download the binaries/XCFramework if they aren't already present and then process them. For android, this wouldn't be as simple. I can cover why in a future post but just know this decision was made with thought behind it. This is also why my CI cannot test android.
See build.gradle and the build-android.sh script to see more on how this works.

## Windows/Linux
Thankfully, windows and linux are super super simple because they both use CMake and are fairly similar. Take a look at the CMakeLists.txt in windows/linux (they differ slightly). Also, see the build-other.sh script for how to cross compile to either platform and archive the results.
Note: Windows is only supported for the windows-x64 target right now, so you only *need* to build an x86_64 target for windows.

Sorry this write up is kind of brief/raw thoughts, but it should be enough to get you started. I am a bit time constrained right now, but maybe if there is interest I could PR this repo and add some documentation about the entire process there, because I think it could be super helpful so nobody would have to go thru the hours of platform-specific debugging and research I did...

**Reply by dinbtechit** [reply]
@GregoryConrad  - omg! A massive thank you.  This is very insightful!!!!! 🙏🙏🙏🙏

> P.S You should totally write that novel hah.  Also, that's an interesting database you got going there. Typo text search wow!  🙌

**Reply by GregoryConrad** [reply]
@dinbtechit I'll be working on https://github.com/fzyzcjy/flutter_rust_bridge/pull/899 over the next several days and should have it done within a week or two. Should help you and others to get started.

**Reply by dinbtechit** [reply]
This is going to be super helpful!!! Thank you soo much.  I am looking forward to it!!

**Comment by GregoryConrad** [comment]
Once #899 is merged, documentation should be available on this in the [user guide](https://cjycode.com/flutter_rust_bridge/library.html)!

**Reply by dinbtechit** [reply]
This documentation is a lifesafer. Found some minor issues tho.

In https://cjycode.com/flutter_rust_bridge/library/dart_only.html 

There is a version mismatch - between `pubspec.yaml` and `cargo.toml`. This is causing compile errors in generated dart code, `bridge_generated.io.dart` unable to resolve type  - `WireSyncReturnStrut`. This can be fixed by upgrading the frb dependency to `1.61.*` in `cargo.toml`. However, we will also need to update the `build.rs` file as  described below.

   ```yaml
   flutter_rust_bridge: ^1.61.0
   ```

   ```toml
    [build-dependencies]
    flutter_rust_bridge_codegen = "1.56.*"        <<---<< Update this to 1.61.*

    [dependencies]
    flutter_rust_bridge = "1.56.*"       <<---<< Update this to 1.61.*
   ```

To fix the above issues update the following in the initiation script ->  [documentation](https://cjycode.com/flutter_rust_bridge/library/dart_only.html ).

1. The following line in the script to fix the `build.rs` file: 
    ```sh
    let raw_opts = RawOpts {
        rust_input: vec![RUST_INPUT.to_string()],
        dart_output: vec![DART_OUTPUT.to_string()],
        c_output: Some(vec![IOS_C_OUTPUT.to_string(), MACOS_C_OUTPUT.to_string()]),     <<---<<  replace this 
        c_output: Some(vec![IOS_C_OUTPUT.to_string()]),            <<---<<  with this 
        inline_rust: true,
        wasm: true,
        ..Default::default()
    };
    ```
2.  And update the versions to `1.61.*`
     
     ```sh
      [lib]
      crate-type = ["staticlib", "cdylib"]

      [build-dependencies]
      flutter_rust_bridge_codegen = "1.61.*"

      [dependencies]
      flutter_rust_bridge = "1.61.*"
      EOF
     ``` 
Now you can do `cargo build` without any errors.

**Reply by GregoryConrad** [reply]
A PR with those changes would be welcome! One thing to note: you won't be generating the C header for macOS anymore; you will need something like this at the bottom of `build.rs`:

```rust
    // Copy ios/ generated C files to macos/
    std::fs::copy(IOS_C_OUTPUT, MACOS_C_OUTPUT).unwrap();
```
